pkg/diff/git.go

```diff
@@ -0,0 +1,451 @@
+			continue
+		}
+		if len(ss) != 3 {
+			log.Printf("[WARN] unknown remote format. remote: %q", s)
+			continue
+		}
+		if ss[2] != "(push)" {
+			continue
+		}
+
+		suffix, found := strings.CutPrefix(ss[1], strings.TrimSuffix(g.url, ".git"))
+		if !found {
+			log.Printf("[WARN] unexpected remote url. expected prefix: %s, actual: %s", strings.TrimSuffix(g.url, ".git"), ss[1])
+			continue
+		}
+
+		cmd, partialfilter, stderr, err := execGitCmd(g.dir, "config", fmt.Sprintf("remote.%s.partialclonefilter", ss[0]))
+		if err != nil && cmd.ProcessState.ExitCode() != 1 {
+			return errors.Wrapf(err, "git config. command: %s, stdout: %s, stderr: %s", cmd.String(), partialfilter, stderr)
+		}
+
+		suffix, found = strings.CutPrefix(suffix, "-archive-")
+		switch found {
+		case false: // not archived but latest remmote
+			g.remotes[ss[0]] = remote{
+				name:         ss[0],
+				url:          ss[1],
+				archiveIndex: 0,
+				partial:      strings.TrimSpace(partialfilter) != "",
+			}
+		case true:
+			index, err := strconv.Atoi(strings.TrimSuffix(suffix, ".git"))
+			if err != nil {
+				log.Printf("[WARN] archive index not found. remote: %s, url: %s", ss[0], ss[1])
+				continue
+			}
+			g.remotes[ss[0]] = remote{
+				name:         ss[0],
+				url:          ss[1],
+				archiveIndex: index,
+				partial:      strings.TrimSpace(partialfilter) != "",
+			}
+		}
+	}
+
+	if g.old != "" {
+		if err := g.ensureCommitExists(g.old); err != nil {
+			return errors.Wrapf(err, "ensure commit %s exists", g.old)
+		}
+	}
+	if g.new != "" {
+		if err := g.ensureCommitExists(g.new); err != nil {
+			return errors.Wrapf(err, "ensure commit %s exists", g.old)
+		}
+	}
+
+	return nil
+}
+
+// ensureCommitExists confirms the existence of commit by "git ls-tree" commmand in tree steps:
+// 1. just run the command
+// 2. if commit not found, fetch all known remotes
+// 3. if still not found, add archive repositories
+func (g gitRepo) ensureCommitExists(commit string) error {
+	cmd, stdout, stderr, err := execGitCmd(g.dir, "ls-tree", "--name-only", commit)
+	if err == nil {
+		return nil
+	}
+	if !strings.Contains(stderr, "Not a valid object name") {
+		return errors.Wrapf(err, "git ls-tree. command: %s, stdout: %s, stderr: %s", cmd.String(), stdout, stderr)
+	}
+
+	if err := g.fetchAll(); err != nil {
+		return errors.Wrapf(err, "fetch all")
+	}
+	cmd, stdout, stderr, err = execGitCmd(g.dir, "ls-tree", "--name-only", commit)
+	if err == nil {
+		return nil
+	}
+	if !strings.Contains(stderr, "Not a valid object name") {
+		return errors.Wrapf(err, "git ls-tree. command: %s, stdout: %s, stderr: %s", cmd.String(), stdout, stderr)
+	}
+
+	if err := g.addArchives(); err != nil {
+		return errors.Wrapf(err, "add archive repositories")
+	}
+	cmd, stdout, stderr, err = execGitCmd(g.dir, "ls-tree", "--name-only", commit)
+	if err != nil {
+		return errors.Wrapf(err, "git ls-tree. command: %s, stdout: %s, stderr: %s", cmd.String(), stdout, stderr)
+	}
+
+	return nil
```

Comment:
```suggestion
	cmd, stdout, stderr, err = execGitCmd(g.dir, "ls-tree", "--name-only", commit)
	if err == nil {
		return nil
	}
	if !strings.Contains(stderr, "Not a valid object name") {
		return errors.Wrapf(err, "git ls-tree. command: %s, stdout: %s, stderr: %s", cmd.String(), stdout, stderr)
	}

	return errors.Errorf("commit: %s is not in remote and archive", commit)
```

Reply:
Reflected in bef03b3 and some (already squashed) commit